### PR TITLE
perf(kernel): Optimize MoE prefill GEMMs for gfx950

### PR DESCRIPTION
**Pull request by Max191**
### Benchmark Summary

Below are comparison tables for different relevant configurations at Po2 sequence lengths from 256-8192 with `topk=4`. The different configurations include:

 - **upstream/main**: Old kernel before this PR
 - **no LLIR no preshuffle**: New kernel without preshuffled weights and the default tokenspeed-triton
 - **no LLIR yes preshuffle**: New kernel with preshuffled weights and the default tokenspeed-triton
 - **yes LLIR yes preshuffle**: New kernel with preshuffled weights and a custom tokenspeed-triton build with LLIR scheduler
 - **AITER (vLLM)**: The AITER kernel (our target benchmark)

The key results to compare are **no LLIR yes preshuffle** and **AITER (vLLM)**. Enabling LLIR can happen later if needed, but is not part of this PR, so the preshuffled version without LLIR is what we get from this PR.

Overall, the optimized kernel beats AITER on `seqlen >= 1024`, and comes close for smaller sequence lengths.

### Dispatch GEMM

| Tokens | upstream/main | no LLIR no preshuffle | no LLIR yes preshuffle | yes LLIR yes preshuffle | AITER (vLLM) |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 256 | 321.723 | 314.483 | 201.722 | 194.682 | 188.162 |
| 512 | 326.523 | 322.683 | 198.323 | 209.682 | 215.682 |
| 1024 | 336.204 | 331.603 | 212.162 | 234.963 | 266.923 |
| 2048 | 471.245 | 457.485 | 325.764 | 318.483 | 420.484 |
| 4096 | 608.526 | 557.806 | 521.806 | 498.245 | 772.288 |
| 8192 | 973.530 | 893.729 | 817.448 | 767.687 | 1455.814 |

### Combine GEMM

| Tokens | upstream/main | no LLIR no preshuffle | no LLIR yes preshuffle | yes LLIR yes preshuffle | AITER (vLLM) |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 256 | 175.162 | 157.841 | 104.921 | 103.081 | 104.001 |
| 512 | 183.721 | 164.081 | 114.201 | 112.281 | 113.441 |
| 1024 | 194.921 | 170.442 | 121.641 | 123.201 | 143.281 |
| 2048 | 276.323 | 230.563 | 165.282 | 164.121 | 200.482 |
| 4096 | 432.364 | 313.604 | 245.842 | 237.683 | 279.003 |
| 8192 | 707.847 | 509.686 | 368.324 | 349.884 | 460.445 |

### Full Fused MoE

| Tokens | upstream/main | no LLIR no preshuffle | no LLIR yes preshuffle | yes LLIR yes preshuffle | AITER (vLLM) |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 256 | 586.446 | 561.126 | 399.884 | 395.164 | 330.324 |
| 512 | 603.726 | 576.566 | 413.084 | 413.084 | 379.164 |
| 1024 | 619.606 | 604.407 | 442.045 | 455.685 | 467.645 |
| 2048 | 860.289 | 792.128 | 599.007 | 594.566 | 693.287 |
| 4096 | 1149.851 | 996.650 | 886.329 | 857.329 | 1137.172 |
| 8192 | 1824.778 | 1539.896 | 1344.654 | 1275.213 | 2042.141 |

### Decode

Since the preshuffled layout is now enabled by default, the decode kernel also gets some basic updates to make it work with the preshuffled layout. This improves performance on the larger M cases:

### topk=2

| M | baseline (us) | new (us) | Speedup % |
|---:|---:|---:|---:|
| 1 | 71.541 | 72.117 | -0.80 |
| 2 | 72.015 | 72.427 | -0.57 |
| 4 | 71.779 | 71.737 | 0.06 |
| 8 | 60.929 | 61.215 | -0.47 |
| 16 | 87.603 | 76.371 | 14.71 |

### topk=4

| M | baseline (us) | new (us) | Speedup % |
|---:|---:|---:|---:|
| 1 | 71.659 | 72.062 | -0.56 |
| 2 | 71.928 | 72.699 | -1.06 |
| 4 | 71.616 | 71.906 | -0.40 |
| 8 | 90.704 | 84.839 | 6.91 |
| 16 | 166.711 | 145.171 | 14.84 |

## Test Plan

The included `test_gluon_moe_gemm_gfx950.py` test was used to verify numerical correctness of the kernel.

Numerics for decode are bit-identical with the new preshuffled layout compared to the non-preshuffled layout. The decode test is not updated to use the preshuffled layout because we have not moved the preshuffling logic into tokenspeed_kernel_amd yet. Once that gets moved (by https://github.com/lightseekorg/tokenspeed/pull/480), then we can add a test for the preshuffled version.

### CODE REVIEW

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py [hunk 1]
```diff
@@ -115,23 +115,22 @@ def _default_num_buffers(
...
+        block_k,
+        block_m,
+        block_n,
+        x_format,
+        w_format,
+        has_x_block_scale,
+        has_w_block_scale,
+        scale_load_mode,
+    )
+    # Match the gfx950 Gluon tutorial v5 local-prefetch pipeline: the
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: "We perform 3-stage pipelining: global -> LDS, LDS -> VGPR, .."

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py [hunk 2]
```diff
@@ -1284,72 +1431,87 @@ def issue_local_loads(self, mfma_idx):
...
+
         return x, w, scale_x, scale_w
 
     @gluon.jit
     def pipeline(self, loop_k):
         cfg = self.cfg
-        EVEN_K: gl.constexpr = cfg.EVEN_K
+        gl.static_assert(
+            cfg.NUM_BUFFERS == 2,
+            "v5 local-prefetch pipeline requires exactly two LDS buffers",
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: "Current pipeline scheme requires .."

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py [hunk 3]
```diff
@@ -2436,33 +2864,59 @@ def _pipelined_moe_tile_compute(
...
         stride_xk,
-        mask_m[:, None],
+        x_mask_nonk,
         k_limit_x,
     )
-    if cfg.W_VIA_VGPR:
-        # Host-preshuffled W -> VGPR direct; bypasses LDS staging.
+    if cfg.W_PRESHUFFLED or cfg.W_VIA_VGPR:
+        # Host-preshuffled W uses the dot-tile HBM order. The current
+        # experiment stages that flat tile through LDS unless W_VIA_VGPR
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: What sure what "the current experiement" means?

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py [hunk 4]
```diff
@@ -2791,6 +3251,30 @@ def _pipelined_moe_tile_compute(
                     block_bases=[],
                     shape=[SUB_BN // 16, BLOCK_K_W * 16],
                 )
+                if cfg.W_PRESHUFFLED and not cfg.W_VIA_VGPR:
```
Comment (antiagainst) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: These if-else conditions in `_pipelined_moe_tile_compute` are going out of control.. Maybe we should factor them out into some utility functions.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py (hunk 3 above)
Comment (Max191) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: This was a stale agent comment. Fixed now.

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py (hunk 4 above)
Comment (Max191) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/moe/fused_mxfp_gfx950.py]: I refactored it into 3 different tile compute functions for: `_run_moe_tile_w_via_vgpr`, `_run_moe_tile_preshuffled_lds_w`, `_run_moe_tile_transposed_w`. I want to do more organization of the whole kernel, but I will leave it as a followup so this PR doesn't get too messy.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py [hunk 5]
```diff
@@ -54,6 +54,76 @@
...
+                    device=w2_scale.device,
+                ),
+            ],
+            dim=-2,
+        ),
+        requires_grad=False,
+    )
+
+
+def _attach_gluon_preshuffle(w: torch.nn.Module) -> None:
```
Comment (borontion) [inline tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:105]: what's the purpose of this function?

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py [hunk 6]
```diff
@@ -54,6 +54,76 @@
...
+            continue
+        if logical_n is not None and int(logical_n) != int(shuffled.shape[-1]):
+            shuffled.original_n = int(logical_n)
+            raw.original_n = int(logical_n)
+            wrapped.original_n = int(logical_n)
+        raw._gluon_shuffled = shuffled
+        wrapped._gluon_shuffled = shuffled
 
 
 def _swizzle_mxfp4(quant_tensor, scale, num_warps):
```
Comment (borontion) [inline tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:129]: I feel there are many kernel specific preshuffling logic here - maybe we should consider moving this swizzle_mxfp4 into tokenspeed-kernel-amd?

and ideally we should test preshuffling there as well instead of tokenspeed-kernel/test/ops/moe/test_gluon_moe_gemm_gfx950.py.

Comment (Max191) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:129]: I have another PR almost ready to send out that does exactly this, but I'm going to land this one first. I'll link the PR here once I send it out.

**Review by borontion — APPROVED** [review]

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py (hunk 5 above)
Comment (Max191) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:105]: This preshuffles the weights into a new tensor with exactly the dot layout that the gluon kernel prefers. Right now it is allocating a new set of weights tensors, which may be problematic. We will need to update the warp_decode implementation to avoid allocating the extra tensor, but I don't want this PR to be too large.

Comment (borontion) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:105]: cc @panditsa

**Review by antiagainst — APPROVED** [review]

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py (hunk 6 above)
Comment (Max191) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/gluon/mxfp4.py:129]: PR link: https://github.com/lightseekorg/tokenspeed/pull/480
